### ISSUE TITLE
having problem to identify porn images -> especaly with penises (NO its not a joke!!!)

### ISSUE THREAD

**Issue by puckpuck85** [tensorflow/tensorflow#11147]
i m running a website where a mass on photos are uploaded (upload without registration)... so i get aaaaa lot of penis-trolls or whatever makes them post their private parts... anyhow...

i tried to train inception model with tensorflow (newest version 1.x) - so i made a   folder with "penises" (approx 160 pics) and one with different images with person how dont show their penis (approx 160 pics).... 

training accuracy is quit good -> over 90 % - but testing the model with other pictures - it fails really bad on detecting penises.... 

hmmmm, i know.. 160 pics are not that much for training, but i thing the problem is:

guy on the beach in shorts is: ok
guy on the beach with penis lurking out of his pants: is not okay.. 

but the difference is quite small between the pictures... because a penis is (mostly) quite a small part of the human body ... so hard to detect... 

anybody could help? and no, its sounds funny, buts no joke.. i suffering under all the uploads i have to review manualy.... it's really not funny to see over 1000 penises a day :(

cheers, puck

**Comment by xdumaine** [comment]
This sounds like a really hard problem.

**Comment by puckpuck85** [comment]
hoho ;)

**Comment by xdumaine** [comment]
Hopefully you'll get some good tips from people in the community.

**Comment by xdumaine** [comment]
I mean tensorflow is really a head of the curve in solving this type of thing.

**Comment by xdumaine** [comment]
Just hoping we'll get a comment from the members of this  repo.

**Comment by puckpuck85** [comment]
hm, trying it again with 1000 pictures per category...

**Comment by puckpuck85** [comment]
btw_ why does python classify_image.py does not classify humans?

**Comment by puckpuck85** [comment]
even with 1000 per category it totally fails on this task

**Comment by xdumaine** [comment]
Growing your sample size should definitely help.

**Comment by xdumaine** [comment]
It should only be a concern if it makes your script run for more than four hours.

**Comment by xdumaine** [comment]
Fuzziness could be a problem, try to make sure your sample images are smooth.

**Comment by xdumaine** [comment]
Also try softer backgrounds to make sure the subject really pops.

**Comment by xdumaine** [comment]
> penis is (mostly) quite a small part of the human body

#humblebrag

**Comment by xdumaine** [comment]
Sorry, I'm really just unloading my ideas here.

**Comment by xdumaine** [comment]
Are your sample images too similar? You have to make sure there's enough deferens between them to train the script.

**Comment by xdumaine** [comment]
If you get stuck though, don't be afraid to pull-out the big guns and use Watson instead.

**Comment by xdumaine** [comment]
Gotta head to bed. I'll get up in the morning and see if you've whipped out a solution.

**Comment by Dim25** [comment]
@puckpuck85 i've a working solution for this problem. give me your email will be happy to share some insights. sorry can't upload it publicly.

**Comment by vade** [comment]
Maybe https://github.com/yahoo/open_nsfw is helpful?

**Comment by jubjamie** [comment]
_Warning: As this doesn't appear to be a bug with Tensorflow, the devs may ask for this to be moved to Stack Overflow._

@vade has posted a good answer for Caffe. Shoutout to @xdumaine for filling the issue tracker with unfunny comments instead of actually helping.

@puckpuck85 You have a couple of ways of attacking this. Your current method of classifying between different image classes has potential but will need much more training data. As you've said, the difference between many of these images can be rather subtle in some instances and more obvious in others. NSFW content can take so many forms that just 160 images won't cover the breadth of the different types of images you are likely to come across. Looking at the repo @vade posted it appears that they used a large dataset but it is private. This was labelled manually of course.

If you want to quickly bolster your dataset you can always scrape images from google images and train on that with inception. I've had surprising success with this! You can then implement a function on your website that if there is a certain uncertainty, it goes to manual review and if it is blocked it gives the user a chance to appeal or whatever. 

Your alternative option (which may be slightly overkill for this situation) is to implement an object detector with annotated objects. Then you can filter out images with have detections. This is a lot of work however as i'm 99% sure that dataset does not already exist.

For the meantime I would suggest increasing your dataset. Perhaps you have a repository of "banned" images from your website that you could use to train? 

I suggest you move this to Stack Overflow as, whilst your question is extremely valid and the basics of image training, it is neither a bug or feature request and i'm not sure how the devs will feel about this on their issue tracker! I'm sure it will make them chuckle.

If you decide to move this issue to Stack Overflow, please post the link here so that myself (or others that stumble upon this issue) can find the S/O thread.

**Comment by xdumaine** [comment]
@jubjamie I made jokes because this is clearly a troll issue. It's a brand new account with 0 activity who's profile picture is [a meme](http://knowyourmeme.com/memes/hide-the-pain-harold). He got you and others to spend time on a fake issue about dick pics. Congrats. You've been bamboozled.

**Comment by jubjamie** [comment]
@xdumaine fair enough. It's not a brand new account and has activity in the area over the last few months. It's still an interesting question and I don't really regret using my saved replies and 2 minutes to reply to it haha. Regardless of context it's still an issue that's seen a number of times both on this issue tracker and stack overflow. I'd rather help someone that might need it than spam the issue tracker with nonsense. 
Will leave this to be closed off. Cheers

**Comment by puckpuck85** [comment]
i'm not trolling at all - caffe with yahoo NSFW works pretty well, but fails sometimes when it comes to close up shots from the penis. 

but, i guess, it will filter 90 % - i will take this route and the rest has to be done by user reporting unwanted pics.... when it works good enough for yahoo - i'm fine.

thread can be closed - i will start the next on stackoverflow. thx all for your help so far.

**Comment by puckpuck85** [comment]
@Dim25 please write me the email to -> puckpuck85@flymail24.com

:)

**Comment by RWOverdijk** [comment]
For now:

https://github.com/infinitered/nsfwjs

**Comment by GantMan** [comment]
Yup!  @RWOverdijk mentioned my TFJS lib

powered by
https://github.com/GantMan/nsfw_model
